lib/propolis/src/attachment.rs

```diff
@@ -0,0 +1,292 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+
+/// Frontend (emulated device) attachment point
+pub struct FrontAttachment<T>(Mutex<AttachPoint<T>>);
+impl<T> FrontAttachment<T> {
+    pub fn new() -> Self {
+        Self(Mutex::new(AttachPoint::default()))
+    }
+    pub fn access<R>(
+        &self,
+        when_attached: impl FnOnce(&T, &AttachState) -> R,
+    ) -> Option<R> {
+        let guard = self.0.lock().unwrap();
+        match &*guard {
+            AttachPoint::Attached(data) => {
+                Some(when_attached(&data.1, &data.0))
+            }
+            AttachPoint::Detached(..) => None,
+        }
+    }
+    pub fn access_fallback<R>(
+        &self,
+        when_attached: impl FnOnce(&T, &AttachState) -> R,
+        fallback: impl FnOnce(&AttachState) -> (),
+    ) -> Option<R> {
+        let guard = self.0.lock().unwrap();
+        match &*guard {
+            AttachPoint::Attached(data) => {
+                Some(when_attached(&data.1, &data.0))
+            }
+            AttachPoint::Detached(state) => {
+                fallback(state);
+                None
+            }
+        }
+    }
+}
+
+/// Backend (resource provider) attachment point
+pub struct BackAttachment<T>(Mutex<AttachPoint<T>>);
+impl<T> BackAttachment<T> {
+    pub fn new() -> Self {
+        Self(Mutex::new(AttachPoint::default()))
+    }
+    pub fn access<R>(
+        &self,
+        when_attached: impl FnOnce(&T, &AttachState) -> R,
+    ) -> Option<R> {
+        let guard = self.0.lock().unwrap();
+        match &*guard {
+            AttachPoint::Attached(data) => {
+                Some(when_attached(&data.1, &data.0))
+            }
+            AttachPoint::Detached(..) => None,
+        }
+    }
+    pub fn access_fallback<R>(
+        &self,
+        when_attached: impl FnOnce(&T, &AttachState) -> R,
+        fallback: impl FnOnce(&AttachState) -> (),
+    ) -> Option<R> {
+        let guard = self.0.lock().unwrap();
+        match &*guard {
+            AttachPoint::Attached(data) => {
+                Some(when_attached(&data.1, &data.0))
+            }
+            AttachPoint::Detached(state) => {
+                fallback(state);
+                None
+            }
+        }
+    }
+}
+
+enum AttachPoint<T> {
+    Attached(Arc<(AttachState, T)>),
+    Detached(AttachState),
+}
+impl<T> Default for AttachPoint<T> {
+    fn default() -> Self {
+        Self::Detached(AttachState::new())
+    }
+}
+
```

Comment:
nit, take it or leave it: it occurs to me that the `access` and `access_fallback` methods on `FrontAttachment` and `BackAttachment` have the same shape, and could be defined in terms of a single shared implementation on `AttachPoint`:
```suggestion
impl<T> FrontAttachment<T> {
    pub fn new() -> Self {
        Self(Mutex::new(AttachPoint::default()))
    }
    pub fn access<R>(
        &self,
        when_attached: impl FnOnce(&T, &AttachState) -> R,
    ) -> Option<R> {
        self.0.lock().unwrap().access(when_attached)
    }
    pub fn access_fallback<R>(
        &self,
        when_attached: impl FnOnce(&T, &AttachState) -> R,
        fallback: impl FnOnce(&AttachState) -> (),
    ) -> Option<R> {
        self.0.lock().unwrap().access_fallback(when_attached, fallback)
    }
}

/// Backend (resource provider) attachment point
pub struct BackAttachment<T>(Mutex<AttachPoint<T>>);
impl<T> BackAttachment<T> {
    pub fn new() -> Self {
        Self(Mutex::new(AttachPoint::default()))
    }
    pub fn access<R>(
        &self,
        when_attached: impl FnOnce(&T, &AttachState) -> R,
    ) -> Option<R> {
        self.0.lock().unwrap().access(when_attached)
    }
    pub fn access_fallback<R>(
        &self,
        when_attached: impl FnOnce(&T, &AttachState) -> R,
        fallback: impl FnOnce(&AttachState) -> (),
    ) -> Option<R> {
        self.0.lock().unwrap().access_fallback(when_attached, fallback)
    }
}

enum AttachPoint<T> {
    Attached(Arc<(AttachState, T)>),
    Detached(AttachState),
}
impl<T> Default for AttachPoint<T> {
    fn default() -> Self {
        Self::Detached(AttachState::new())
    }
}
impl<T> AttachPoint<T> {
    fn access<R>(
        &self,
        when_attached: impl FnOnce(&T, &AttachState) -> R,
    ) -> Option<R> {
        match self {
            AttachPoint::Attached(data) => {
                Some(when_attached(&data.1, &data.0))
            }
            AttachPoint::Detached(..) => None,
        }
    }
    fn access_fallback<R>(
        &self,
        when_attached: impl FnOnce(&T, &AttachState) -> R,
        fallback: impl FnOnce(&AttachState) -> (),
    ) -> Option<R> {
        match self {
            AttachPoint::Attached(data) => {
                Some(when_attached(&data.1, &data.0))
            }
            AttachPoint::Detached(state) => {
                fallback(state);
                None
            }
        }
    }
}
```

It's up to you whether you think the de-duplication is worth it, though.

Reply:
I did some combining here.  With the consolidation into `AttachPoint`, it does seem a bit silly to have `FrontAttachment` and `BackAttachment` structs which are identical save for their names.  That being said, it's probably not worth the simplicity of a single attachment type if we lose the compiler protection from attaching two frontends or two backends together.